update: custom scene management and InstantiateAndSpawn [MTT-7399] (#1142)
Unity-Technologies/com.unity.multiplayer.docs: docs/basics/object-spawning.md

[Learn more about Netcode Scene Management here](scenemanagement/scene-management-overview.md)

> :::caution You might find it useful to add a `GameObject` property in a `NetworkBehaviour` derived component that you use to assign a network prefab instance for dynamically spawning. You need to make sure to instantiate a new instance **prior** to spawning. If you attempt to just spawn the actual network prefab instance it can result in unexpected results.

Reviewer: ```suggestion
:::caution You might find it useful to add a `GameObject` property in a `NetworkBehaviour`-derived component to use when assigning a network prefab instance for dynamically spawning. You need to make sure to instantiate a new instance **prior** to spawning. If you attempt to just spawn the actual network prefab instance it can result in unexpected results.
```

:::

### Taking Prefab Overrides Into Consideration

> Sometimes you might want to make a simpler prefab instance to be spawned on server version the override for clients. You should take this into consideration when dynamically spawning a network prefab. If you are running as a host, you want the override to spawn since a host is both a server and a client. However, if you also want to have the ability to run as a dedicated server you might want to spawn the source network prefab. 
> 
> **_There are two ways you can accomplish this:_**
> 

Reviewer: ```suggestion
Sometimes, you might want to make a simpler prefab instance to be spawned on server version the override for clients. You should take this into consideration when dynamically spawning a network prefab. If you're running as a host, you want the override to spawn since a host is both a server and a client. However, if you also want to have the ability to run as a dedicated server, you might want to spawn the source network prefab. 

There are two ways you can accomplish this, as explained below.

```

#### Get The Network Prefab Override First 

> This option provides you with the over-all view of getting the network prefab override, instantiating it, and then spawning it.

Reviewer: ```suggestion
This option provides you with the overall view of getting the network prefab override, instantiating it, and then spawning it.
```

```csharp
var instance = Instantiate(NetworkManager.GetNetworkPrefabOverride(myPrefab));
var instanceNetworkObject = instance.GetComponent<NetworkObject>();
instanceNetworkObject.Spawn();
```

>  In the above script, you can see that we get the prefab override using the `NetworkManager.GetNetworkPrefabOverride` method, we then create an instance of the network prefab override, and finally we spawn the network prefab override instance's `NetworkObject`. Of course, this is a bit of code to write in order to just "spawn something".

Reviewer: ```suggestion
In the above script, we get the prefab override using the `NetworkManager.GetNetworkPrefabOverride` method. Then we then create an instance of the network prefab override, and finally we spawn the network prefab override instance's `NetworkObject`.
```

> This option provides you with the over-all view of getting the network prefab override, instantiating it, and then spawning it.
> 
> ```csharp
> var instance = Instantiate(NetworkManager.GetNetworkPrefabOverride(myPrefab));
> var instanceNetworkObject = instance.GetComponent<NetworkObject>();
> instanceNetworkObject.Spawn();
> ```
>  In the above script, you can see that we get the prefab override using the `NetworkManager.GetNetworkPrefabOverride` method, we then create an instance of the network prefab override, and finally we spawn the network prefab override instance's `NetworkObject`. Of course, this is a bit of code to write in order to just "spawn something".
> 

Reviewer: ```suggestion
This option provides you with the overall view of getting the network prefab override, instantiating it, and then spawning it.

```csharp
var instance = Instantiate(NetworkManager.GetNetworkPrefabOverride(myPrefab));
var instanceNetworkObject = instance.GetComponent<NetworkObject>();
instanceNetworkObject.Spawn();
```
 In the above script, we get the prefab override using the `NetworkManager.GetNetworkPrefabOverride` method. Then we then create an instance of the network prefab override, and finally we spawn the network prefab override instance's `NetworkObject`.

```

#### Using InstantiateAndSpawn

> The second option you have is to leverage from the `NetworkSpawnManager.InstantiateAndSpawn` method that handles dealing with whether to spawn an override or not for you. The below script was written as if was being invoked within a `NetworkBehaviour`. 

Reviewer: ```suggestion
The second option is to leverage the `NetworkSpawnManager.InstantiateAndSpawn` method that handles dealing with whether to spawn an override or not for you. The below script was written as if it was being invoked within a `NetworkBehaviour`. 
```

```csharp
var networkObject = NetworkManager.SpawnManager.InstantiateAndSpawn(myPrefab, ownerId);
```

> We pass in the overridden source network prefab we want to have instantiated and spawned, and then it returns the instantiated and spawned `NetworkOject` of the spawned object. The default behavior of `InstantiateAndSpawn` is to spawn the override if running as a host and the original source prefab if running as a server.
> 
> `InstantiateAndSpawn` has several more parameters to provide a bit more control over this process:

Reviewer: ```suggestion
We pass in the overridden source network prefab we want to have instantiated and spawned, and then it returns the instantiated and spawned `NetworkObject` of the spawned object. The default behavior of `InstantiateAndSpawn` is to spawn the override if running as a host and the original source prefab if running as a server.

`InstantiateAndSpawn` has several parameters to provide more control over this process:
```

```csharp
InstantiateAndSpawn(NetworkObject networkPrefab, ulong ownerClientId = NetworkManager.ServerClientId, bool destroyWithScene = false, bool isPlayerObject = false, bool forceOverride = false, Vector3 position = default, Quaternion rotation = default)
```

> Looking at the parameters, we can see it defaults to the server as being the owner, assures the instantiated `NetworkObject` will not be destroyed if the scene is unloaded, is not spawned as a player, has a "forceOverride" parameter (more on this in a second), and provides a way to set the position and rotation of the newly instantiated `NetworkObject`.
> 
> The "forceOverride" parameter, when set to true, will use the override whether you are running a server or host.

Reviewer: ```suggestion
Looking at the parameters, we can see it defaults to the server as the owner, ensures that the instantiated `NetworkObject` won't be destroyed if the scene is unloaded, is not spawned as a player, has a `forceOverride` parameter, and provides a way to set the position and rotation of the newly instantiated `NetworkObject`.

The `forceOverride parameter`, when set to true, will use the override whether you're running as either server or host.
```

> The second option you have is to leverage from the `NetworkSpawnManager.InstantiateAndSpawn` method that handles dealing with whether to spawn an override or not for you. The below script was written as if was being invoked within a `NetworkBehaviour`. 
> 
> ```csharp
> var networkObject = NetworkManager.SpawnManager.InstantiateAndSpawn(myPrefab, ownerId);
> ```
> We pass in the overridden source network prefab we want to have instantiated and spawned, and then it returns the instantiated and spawned `NetworkOject` of the spawned object. The default behavior of `InstantiateAndSpawn` is to spawn the override if running as a host and the original source prefab if running as a server.
> 
> `InstantiateAndSpawn` has several more parameters to provide a bit more control over this process:
> 
> ```csharp
> InstantiateAndSpawn(NetworkObject networkPrefab, ulong ownerClientId = NetworkManager.ServerClientId, bool destroyWithScene = false, bool isPlayerObject = false, bool forceOverride = false, Vector3 position = default, Quaternion rotation = default)
> ```
> 
> Looking at the parameters, we can see it defaults to the server as being the owner, assures the instantiated `NetworkObject` will not be destroyed if the scene is unloaded, is not spawned as a player, has a "forceOverride" parameter (more on this in a second), and provides a way to set the position and rotation of the newly instantiated `NetworkObject`.
> 
> The "forceOverride" parameter, when set to true, will use the override whether you are running a server or host.
> 
> 

Reviewer: ```suggestion
The second option is to leverage the `NetworkSpawnManager.InstantiateAndSpawn` method that handles dealing with whether to spawn an override or not for you. The below script was written as if it was being invoked within a `NetworkBehaviour`. 

```csharp
var networkObject = NetworkManager.SpawnManager.InstantiateAndSpawn(myPrefab, ownerId);
```
We pass in the overridden source network prefab we want to have instantiated and spawned, and then it returns the instantiated and spawned `NetworkObject` of the spawned object. The default behavior of `InstantiateAndSpawn` is to spawn the override if running as a host and the original source prefab if running as a server.

`InstantiateAndSpawn` has several more parameters to provide a bit more control over this process:

```csharp
InstantiateAndSpawn(NetworkObject networkPrefab, ulong ownerClientId = NetworkManager.ServerClientId, bool destroyWithScene = false, bool isPlayerObject = false, bool forceOverride = false, Vector3 position = default, Quaternion rotation = default)
```

Looking at the parameters, we can see it defaults to the server as the owner, assures the instantiated `NetworkObject` won't be destroyed if the scene is unloaded, is not spawned as a player, has a "forceOverride" parameter, and provides a way to set the position and rotation of the newly instantiated `NetworkObject`.

The `forceOverride` parameter, when set to true, will use the override whether you are running a server or a host.

```
